src/vulkan_state.h

```diff
@@ -22,18 +22,42 @@
 
 #pragma once
 
+#include <functional>
 #include <vulkan/vulkan.hpp>
 
 #include "managed_resource.h"
+#include "vulkan_wsi.h"
 
-class VulkanWSI;
 
 class VulkanState
 {
 public:
-    VulkanState(VulkanWSI& vulkan_wsi);
+    using ChoosePhysicalDeviceStrategy = 
+        std::function<vk::PhysicalDevice (std::vector<vk::PhysicalDevice> const&)>;
+    
+    VulkanState(VulkanWSI& vulkan_wsi, ChoosePhysicalDeviceStrategy const& pd_strategy)
+    {
+        create_instance(vulkan_wsi);
+        create_physical_device(vulkan_wsi, pd_strategy);
+        create_logical_device(vulkan_wsi);
+        create_command_pool();
+    }
+
+    void create_physical_device(VulkanWSI& vulkan_wsi, ChoosePhysicalDeviceStrategy const& pd_strategy)
+    {
+        vk_physical_device = pd_strategy(available_devices(vulkan_wsi));
+    }
 
-    void log_info();
+    void log_info()
+    {
+        log_info(physical_device());
+    }
+
+    void log_all_devices()
+    {
+        // all devices not devices supported by wsi
+        log_all_devices(instance().enumeratePhysicalDevices());
+    }
```

Comment:
Similarly, I am not sure what's the benefit of exposing the internals of log_* and also having redirects to private versions of log_* methods.

An alternative would be to have a log_device_info() function internal to the .cpp file (i.e., not even a member of VulkanState). log_info() would just call that, whereas log_all_devices() would also directly call that in a loop. Basically do what the private log_* currently do but without a need to accept any arguments.

Reply:
That is fine by me. I didn't like idea of function doing two jobs (deciding which device/devices to write and how to write the message). For how to write i had functions yesterday because i believe it is not the vulkan_state responsibility. I probably misunderstood your yesterdays comment and made them member functions, and the choosing part was those exposed ones.

Now it is like the middle-ground i really like. Free function and hidden control methods. Thank you for your idea